### ISSUE TITLE
IMagick extension version check does not work as expected

### ISSUE THREAD

**Issue by pun-ky** [dompdf/dompdf#1296]
see https://github.com/avalanche123/Imagine/issues/474

please make some defensive code which will handle that bug, Currently I have to manually hardcode that :/ 

As a workaround, I am just changing:

`$imagickClonable = version_compare(phpversion('imagick'), '3.0.1rc1') > 0;`

to

`$imagickClonable = true; // version_compare(phpversion('imagick'), '3.0.1rc1') > 0;`

**Comment by bsweeney** [comment]
Also relevant is #459.

**Comment by bsweeney** [comment]
The _extension_ removed the clone keyword in version 3.1 (to clarify my statement from #459).

So if I am understanding this right, `phpversion` does not always return the correct value (?). It looks like we should check the extension version using the `IMAGICK_EXTVER` contstant (which is available starting with 2.0.1 so ... should be safe).

Can you try replacing that line with:

```
$imagickClonable = version_compare(IMAGICK_EXTVER, '3.0.1rc1') > 0;
```

**Comment by pun-ky** [comment]
yes, `phpversion('imagick')` is returning incorrect value due to invalid build of that library distributed on some operating systems

**Comment by bsweeney** [comment]
Do you know if that modification will work as expected?

**Comment by bsweeney** [comment]
The addition of the `IMAGICK_EXTVER` constant was noted in the changelog for [v2.0.1](https://pecl.php.net/package-info.php?package=imagick&version=2.0.1).

The deprecation of the `clone()` method was noted in the changelog for [v3.1.0b1](https://pecl.php.net/package-info.php?package=imagick&version=3.1.0b1).

**Comment by bsweeney** [comment]
The updated check is in the code. If anyone still has issues please post back here.

**Comment by bluefirex** [comment]
Using the bleeding edge version of DomPDF (master-branch) I still get `Imagick::clone method is deprecated and it's use should be avoided`.

PHP 7.0.15
Using Imagick (Extension) on macOS (`imagick module version => @PACKAGE_VERSION@`)

**Comment by bsweeney** [comment]
😭

**Comment by bsweeney** [comment]
I think maybe we should just check for the existence of the clone method rather than do a version compare. I believe that how you get the extension may determine whether or not the version constant is populated.

@bluefirex are you able to do a quick test for me? Can you replace [Cpdf.php#L4793](https://github.com/dompdf/dompdf/blob/75b9aa1056c5c4b56cbbdc747a4cce1c10d5c89f/lib/Cpdf.php#L4793) with the following:

```
$imagickClonable = method_exists('\Imagick','clone') == false;
```

That's a little awkward. If that works I may tweak the logic a bit to make it easier to read.

**Comment by bluefirex** [comment]
@bsweeney That doesn't work. The clone method does exist (it couldn't throw an error otherwise).

**Comment by bsweeney** [comment]
right ... sorry ... deprecated

any thoughts?

**Comment by bsweeney** [comment]
(I really need to set this up so I can test.)

**Comment by bluefirex** [comment]
I have an idea: Why not check whether the `__clone` method exists and than use the keyword instead of the method, if it does?

**Comment by bluefirex** [comment]
Okay, apparently Imagick doesn't declare `__clone`. Neither `method_exists()` nor `ReflectionClass::getMethods` can find it.

**Comment by simonberger** [comment]
Doesn't it make sense to find out why the version check fails before we try to find a more obscure workaround??
Maybe the ```IMAGICK_EXTVER``` isn't set on his system, the version to compare to doesn't work with the RC1 addition or whatever.

**Comment by bsweeney** [comment]
That's a good point. `version_compare` _should_ take into account RC versions so I don't think that's the problem.

@bluefirex what does `Imagick::IMAGICK_EXTVER` return on your system.

**Comment by bluefirex** [comment]
`Imagick::IMAGICK_EXTVER` is `3.4.3`

**Comment by bsweeney** [comment]
Hm. And what about that `version_compare` statement?

`version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1')`

**Comment by bsweeney** [comment]
Maybe we should try adding in the operator instead of checking if it's `> 0`?

```
version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1', '>');
```

**Comment by bluefirex** [comment]
`version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1', '>') === true` for me

**Comment by simonberger** [comment]
And the retest with version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1') > 0 really fails? That would be odd. If that works too the problem would be somewhere else.

**Comment by simonberger** [comment]
Haha

By default, version_compare() returns -1 if the first version is lower than the second, 0 if they are equal, and 1 if the second is lower.

**Comment by bsweeney** [comment]
Still seems odd to me, but I think we'll use the operator version (yes?).

**Comment by simonberger** [comment]
I personally don't believe in those fixes that are making no sense. If that would change something it would be a bug in the PHP implementation.
@bluefirex 
Did you check the whole process with the change to 
```version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1', '>') === true```
and could you please verify that the condition 
```version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1') > 0``` really failed ?

**Comment by bluefirex** [comment]
`version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1') > 0 == true` for me so same result as the other condition :/ I didn't change it in the domPDF code, just ran it in a separate file.

**Comment by bluefirex** [comment]
Okay, this is interesting. I'm generating PDFs using domPDF with a PHP script that runs in fpm. In CLI the condition works perfectly fine but in fpm it failed. I just now restarted php-fpm and suddenly it works…

**Comment by bsweeney** [comment]
So we're ok then! I...guess? I think for now maybe this is fodder for a FAQ, unless/until we determine that this really isn't going to work or we find a better way to determine if the version of the extension has deprecated the clone method.

**Comment by samir-plusb** [comment]
Hello,

I am facing the same problem and just wanted to leave this issue here:
https://github.com/mkoppanen/imagick/issues/194

In my case `Imagick::IMAGICK_EXTVER` is always returning the string `@PACKAGE_VERSION@`  which is why the following expression will always be `false`, even if the (compiled) version of Imagick I am using should be correct:

`version_compare(Imagick::IMAGICK_EXTVER, '3.0.1rc1') > 0;`

So basically this problem depends on how the Imagick extension was installed...

**Comment by bsweeney** [comment]
So ... what is the `PHP_IMAGICK_EXTNUM` value based on? Maybe we should use that instead?

**Comment by adam1010** [comment]
I'm having the same issue using the official Windows 7.2 dll from pecl. The module version is not set and is causing other libraries to fail when they try to check the version.

When I run phpinfo(); I see:

imagick module version | @PACKAGE_VERSION@
Imagick compiled with ImageMagick version | ImageMagick 7.0.7-11 Q16 x64 2017-11-23

http://windows.php.net/downloads/pecl/snaps/imagick/3.4.3/php_imagick-3.4.3-7.2-ts-vc15-x64.zip
https://mlocati.github.io/articles/php-windows-imagick.html

Is there a way to force the module to report "3.4.3" or do I have to try and compile this extension myself?

**Comment by reichwebconsulting** [comment]
I've been bashing my head against this problem all night, which seems to be caused by the fact that the precompiled windows binaries don't correctly set Imagick::IMAGICK_EXTVER but instead have the placeholder @PACKAGE_VERSION@.

I got my code to work by replacing line `4860` in `Cpdf.php` with:

`$imagickClonable = (new ReflectionClass(Imagick::class))->isCloneable();`

I hate dropping in reflection because it can be an expensive solution, but it works (at least for me)

**Comment by bsweeney** [comment]
@reichwebconsulting thanks for the update. Of course you're right, but this may be the best fallback. Maybe checking to see if the value is non-semantic (i.e. "@PACKAGE_VERSION") and if it is use reflection.

**Comment by OmathiasO** [comment]
> @reichwebconsulting thanks for the update. Of course you're right, but this may be the best fallback. Maybe checking to see if the value is non-semantic (i.e. "@PACKAGE_VERSION") and if it is use reflection.

I would like to write a Pull Request for that with the combined solution provided by @bsweeney and @reichwebconsulting  if that's okay with you. I just stumbled upon the same Bug (@PACKAGE_VERSION@) and manually correcting the code is not an option due to automated Deployment Pipeline.

**Comment by bsweeney** [comment]
@mathiasonea please do, all pull requests are welcome!

**Comment by 4UForever** [comment]
I have updated my imagick version follows this quote and works., thanks.

> 
> 
> I'm having the same issue using the official Windows 7.2 dll from pecl. The module version is not set and is causing other libraries to fail when they try to check the version.
> 
> When I run phpinfo(); I see:
> 
> imagick module version | @PACKAGE_VERSION@
> Imagick compiled with ImageMagick version | ImageMagick 7.0.7-11 Q16 x64 2017-11-23
> 
> http://windows.php.net/downloads/pecl/snaps/imagick/3.4.3/php_imagick-3.4.3-7.2-ts-vc15-x64.zip
> https://mlocati.github.io/articles/php-windows-imagick.html
> 
> Is there a way to force the module to report "3.4.3" or do I have to try and compile this extension myself?

**Comment by andreaspabst** [comment]
> I have updated my imagick version follows this quote and works., thanks.
> 
> > I'm having the same issue using the official Windows 7.2 dll from pecl. The module version is not set and is causing other libraries to fail when they try to check the version.
> > When I run phpinfo(); I see:
> > imagick module version | @PACKAGE_VERSION@
> > Imagick compiled with ImageMagick version | ImageMagick 7.0.7-11 Q16 x64 2017-11-23
> > http://windows.php.net/downloads/pecl/snaps/imagick/3.4.3/php_imagick-3.4.3-7.2-ts-vc15-x64.zip
> > https://mlocati.github.io/articles/php-windows-imagick.html
> > Is there a way to force the module to report "3.4.3" or do I have to try and compile this extension myself?

I can reproduce it

```
php -a
Interactive shell

php > var_dump(Imagick::getVersion());
array(2) {
  ["versionNumber"]=>
  int(1802)
  ["versionString"]=>
  string(67) "ImageMagick 7.0.10-57 Q16 x86_64 2021-01-10 https://imagemagick.org"
}
php > var_dump(Imagick::IMAGICK_EXTVER);
string(17) "@PACKAGE_VERSION@"
```

**Comment by emtudo** [comment]
I'm having the same problem, using php 8.

ImageMagick 7.0.10-48.

(I am unable to update the version of ImageMagic at this time).

**Comment by bsweeney** [comment]
Perhaps it's time to remove the conditional check (#2373)? Support for the `clone` keyword in the extension was released more than 10 years ago. Would need to add a note indicating the supported version to composer suggest.

And to avoid potentially breaking with older installations I'd also make usage of the extension optional and disabled by default.

**Comment by adam1010** [comment]
The `imagick` extension for php8 isn't on PEAR yet so we're compiling from source -- when doing so there is no version number set and domPDF's version check fails.  This PR fixes it:  https://github.com/dompdf/dompdf/pull/2373

**Comment by nicoschreiner** [comment]
I do have the same problem, that the version is set to `@PACKAGE_VERSION@`. I'm using [Bref.sh](https://bref.sh/) (PHP 7.4) with the imagick extension from [brefphp/extra-php-extensions](https://github.com/brefphp/extra-php-extensions). Sadly I can't easily change the imagick version which is used.

The PR (#2373) mentioned by @adam1010 looks promising. Does anybody know when (and if) this will be merged?

**Comment by bsweeney** [comment]
Obviously a change is necessary here, but if I implement the suggested change how many issues will be opened about imagick integration no longer working because there are still installations of older packages? It's too bad you can't add a version requirement to suggest section in Composer.

Since I'm trying to be more cognizant about breaking changes we'll switch to always using the clone keyword for the next major release. For the next minor release, instead of outright removing the check we can do something like the following:

```
static $imagickClonable = null;
if ($imagickClonable === null) {
    $imagickClonable = true;
    if (defined('Imagick::IMAGICK_EXTVER')) {
        $imagickVersion = Imagick::IMAGICK_EXTVER;
    } else {
        $imagickVersion = 0;
    }
    if (version_compare( $imagickVersion, '0.0.1', '>=' ) >= 0) {
        $imagickClonable = version_compare(\Imagick::IMAGICK_EXTVER, '3.0.1rc1') > 0;
    }
}
```

**Comment by hebinet** [comment]
> Obviously a change is necessary here, but if I implement the suggested change how many issues will be opened about imagick integration no longer working because there are still installations of older packages? It's too bad you can't add a version requirement to suggest section in Composer.
> 
> Since I'm trying to be more cognizant about breaking changes we'll switch to always using the clone keyword for the next major release. For the next minor release, instead of outright removing the check we can do something like the following:
> 
> ```
> static $imagickClonable = null;
> if ($imagickClonable === null) {
>     $imagickClonable = true;
>     if (defined('Imagick::IMAGICK_EXTVER')) {
>         $imagickVersion = Imagick::IMAGICK_EXTVER;
>     } else {
>         $imagickVersion = '0';
>     }
>     if (version_compare($imagickVersion, '0.0.1', '>=')) {
>         $imagickClonable = version_compare($imagickVersion, '3.0.1rc1', '>=');
>     }
> }
> ```

Hello,
this solution looks good to me.
Is there any plan to integrate this solution into the package?

**Comment by bsweeney** [comment]
Yes. Things have been ... not optimal ... for me lately. But I do plan to push this out.

**Comment by Alex-DA** [comment]
Hello, maybe this can help you.

Using the official Docker Image of YII2:apache:php-8.0 https://github.com/yiisoft/yii2-docker/blob/master/php/Dockerfile-debian , this is what you get from the ImagickVersion:

```
var_dump(constant("Imagick::IMAGICK_EXTVER"));
string(52) "git--master-132a11fd26675db9eb9f0e9a3e2887c161875206"
```

**Comment by bsweeney** [comment]
Anyone here tried with the latest release yet? Curious if this is still an issue after the change to the logic.

**Comment by nicoschreiner** [comment]
@bsweeney It's working for me now.
